MoD/MoD.py

```diff
@@ -109,26 +113,54 @@ def forward(self, x: torch.Tensor):
         # [Shape] output: (batch_size, sequence_length, hidden_size)
         output = x.clone()
 
-        # Assure that the auxiliary decision is a boolean tensor.
-        aux_decision = aux_decision.bool()
+        # In the train stage, we use the real top-k decision.
+        # In the eval stage, we use the auxiliary router prediction decision.
+        topk_decision = (
+            aux_targets.bool() if self.training else aux_decision.bool()
+        )
 
+        # TODO[keli]: How to enable batch processing for the following loop?
         for b in range(B):
             # Extract tokens and router that need to go through the transformer block.
-            # [Shape] selected_tokens_emb: (selected_tokens_count, hidden_size)
-            selected_tokens_emb = x[b, aux_decision[b]]
+            # `unsqueeze(0)` is used to add the batch dimension back.
```

Comment:
Maybe unnecessary.

Reply:
Great, but I think the code is suck, we can update the whole code later.